### ISSUE TITLE
[Feature request] Support for AC4 and TrueHD?

### ISSUE THREAD

**Issue by sclsj** [VoidXH/Cavern#55]
I can't find a player for AC4 or TrueHD*. I'm still in the process of requesting a trial to the Dolby Reference Player but I haven't heard back for a while (2+ weeks) so it's probably ignored.

Or will this project be for eac3 only?

*Full object support (?): I'm told that TrueHD stores the clustered objects separately (so lossy clustering->lossless (no JOC) ->lossless encode), unlike EAC3 where it's combined into 5.1/7.1 channels using JOC (so lossy clustering->lossy JOC->lossy encode), but mediainfo and ffmpeg all give me 7.1 channels for TrueHD, so I'm not sure what's happening. Maybe TrueHD is already fully supported by FFmpeg (and therefore most of the players).

**Comment by VoidXH** [comment]
This project is for spatial codecs that are properly documented, actually used, and not yet supported by anything else. AC-4, while being completely documented, is not yet used nor supported by commercial hardware. It will stay this way for a long time, as it's using the same JOC tables as AC-3 (the much needed mixing quality improvement is not in that codec), also, AC-3 is completely underutilized right now, it still has breathing room if better quality is needed.

TrueHD is completely undocumented and will likely stay that way. AC-3 had to be an open standard to be used in broadcast, that's the only reason we have docs for it.

**Comment by sclsj** [comment]
Ah, I thought AC4 is much better than EAC3 since it only uses half of the bitrate for the same quality(as claimed by Dolby). What do you mean by "much needed mixing quality improvement"? Are you referring to JOC?

With regards to TrueHD, I seems to be able to open it in FFmpeg / IINA / VLC etc. How do I know if they fully implemented TrueHD decoding and rendering? And does it use JOC also?

By the way, I think I have asked this before, but do you plan on supporting mpeg-h 3d in the future? It's documented and not yet supported by anything else but I'm not sure how widespread / utilized it is.

**Comment by VoidXH** [comment]
> What do you mean by "much needed mixing quality improvement"? Are you referring to JOC?

Yes. JOC can only mix from 10 subbands per input channel, and this is the issue, not precision. The subbands are the exact same for AC-4.

> How do I know if they fully implemented TrueHD decoding and rendering?

The commit history (of mlpdec.c) and comments help. Currently, FFmpeg just throws away the Atmos metadata. The fact that TrueHD is supported at all is a huge achievement, since it had to be reverse engineered.

> And does it use JOC also?

It is speculated from the file sizes that it is indeed JOC. Without proper documentation, nobody can be sure.

> do you plan on supporting mpeg-h 3d in the future?

libmpegh supports it completely.

**Comment by VoidXH** [comment]
> I'm not that familiar with JOC. Could you please explain more?

JOC is a matrix mixer on steroids. Its inputs are not the source channels (5.0 or 7.0), but multiple subbands of them. An input channel is disassembled to multiple parts like front left between 100-200 Hz and front left between 200-300 Hz, etc. These are not real examples, subband does not exactly mean a band limit, but a part of a channel in some other way. JOC only contains the ratio of these subbands. Any output object can contain any channel's any subband in any gain. While this looks like many possibilities, it's still only 10 subbands, that's not too many choices to LEGO the result together.

> Is there a significance of Atmos metadata other than object decorrelation / rendering

Objects are required if you want to go beyond 7.1. Channel-based mixes are also contained in objects, even though the codec supports completely channel-based 9.1.6, without object metadata or JOC, but nobody is using it.

> Artists also set way too aggressive gain

Sounds like a decoder issue. I had something like this for StormAudio, I had to export all my AC-3 files at -3 dB, otherwise it ran into clipping on the processor.

> Are you talking about EAC-3?

Yes. It has features like advanced hybrid transform, spectral extensions, positional object metadata with extended precision, but all of these are completely unused currently.

> Are we still stuck at only one extension packet?

This is not the official naming for anything. An E-AC-3 stream consists of substreams, up to 8 of both types (E-AC-3 or regular AC-3, totalling at 16). This means that even 16 channels of lossless mono can be wrapped to AC-3. For compatibility reasons, only the first substream is used for any content. A substream's DWORD count is 11 bits, this means the maximum size per subframe is 16*2^11 bits. A subframe contains 1536 samples, this accounts to 1 Mbit/s bandwidth per subframe. JOC is contained in a previously unused field called skipfld. This is also size limited to 256 kbit/s, and can only be present in one substream.

There's only one case where a second substream is used: 7.1. As E-AC-3 can only support 5 full bandwidth and 1 LFE channels per substream, to have other channels, you will need some hacking. These demos (only found on Dolby Demo Discs) have a dowmixed 5.1 and a 4.0 substream that overrides the sides and adds rears. More substreams can be added to reach 9.1.6, but this is also an unused feature. I'm working on a backwards-compatible nearly lossless E-AC-3 9.1.6 encoder without JOC, hopefully it's supported by others and not just Cavern.

**Comment by ValZapod** [comment]
>The fact that TrueHD is supported at all is a huge achievement, since it had to be reverse engineered.

Yeah, well, it was just broken again!  See this issue https://trac.ffmpeg.org/ticket/10126#comment:4 and there are still some bugs in it compared to reference decoders (be it HW LG C9 or Plex's EAE). https://trac.ffmpeg.org/ticket/9958

**Comment by hockinsk** [comment]
> I can't find a player for AC4 or TrueHD*. I'm still in the process of requesting a trial to the Dolby Reference Player but I haven't heard back for a while (2+ weeks) so it's probably ignored.
> 
> Or will this project be for eac3 only?
> 
> *Full object support (?): I'm told that TrueHD stores the clustered objects separately (so lossy clustering->lossless (no JOC) ->lossless encode), unlike EAC3 where it's combined into 5.1/7.1 channels using JOC (so lossy clustering->lossy JOC->lossy encode), but mediainfo and ffmpeg all give me 7.1 channels for TrueHD, so I'm not sure what's happening. Maybe TrueHD is already fully supported by FFmpeg (and therefore most of the players).

Dolby Reference Player only supports AC4-IMS playback to virtualized headphones and virtualized stereo speakers as experience on Android devices supporting Atmos. AC4-IMS is used by Tidal and according to Dolby  Apple Music for Android users.

**Comment by ValZapod** [comment]
>AC4-IMS is used by Tidal

No. They moved to EAC3 from AC4.

**Comment by sclsj** [comment]
> > AC4-IMS is used by Tidal
> 
> No. They moved to EAC3 from AC4.

I believe both are available. I can download in either format with orpheusdl-tidal.

**Comment by ValZapod** [comment]
>files are in m4a format

They are not, they are mp4.

>like I need to convert it to mp4

m4a is a subset of mp4.

**Comment by sclsj** [comment]
> > files are in m4a format
> 
> They are not, they are mp4.
> 
> > like I need to convert it to mp4
> 
> m4a is a subset of mp4. No convertion needed.

When I download them using orpheus-dl tidal module, they are m4a files. I cannot open them in Dolby Reference Player.

**Comment by ValZapod** [comment]
>module, they are m4a files


mp42/dby1 is mp4, not m4a. This is a kbown bug in tidal-dl.

**Comment by sclsj** [comment]
> > module, they are m4a files
> 
> mp42/dby1 is mp4, not m4a. This is a known bug in tidal-dl.


Yes, you are right. I just renamed it to mp4 and was able to open it. Why didn't I thought of doing that?

**Comment by hockinsk** [comment]
You have to demux from tidal in order to play in Dolby Reference Player.

> > > I can't find a player for AC4 or TrueHD*. I'm still in the process of requesting a trial to the Dolby Reference Player but I haven't heard back for a while (2+ weeks) so it's probably ignored.
> > > Or will this project be for eac3 only?
> > > *Full object support (?): I'm told that TrueHD stores the clustered objects separately (so lossy clustering->lossless (no JOC) ->lossless encode), unlike EAC3 where it's combined into 5.1/7.1 channels using JOC (so lossy clustering->lossy JOC->lossy encode), but mediainfo and ffmpeg all give me 7.1 channels for TrueHD, so I'm not sure what's happening. Maybe TrueHD is already fully supported by FFmpeg (and therefore most of the players).
> > 
> > 
> > Dolby Reference Player only supports AC4-IMS playback to virtualized headphones and virtualized stereo speakers as experienced on Android devices supporting Atmos. AC4-IMS is used by Tidal and according to Dolby Apple Music for Android users.
> 
> Tidal on iOS seems to not be using system Dolby Atmos decoder because the control center shows "stereo". How can I tell if it's actually playing the Dolby Atmos version, and if it is, if it's streaming eac3 or ac4?

Apple devices don't support ac4-ims, only dd+joc so Tidal streams dd+joc to apple and ac4-ims to Android. Dolby say Apple Music stream ac4-ims to Android, but I received dd+joc when I looked into it. Looks like Apple Music encodes the joc differently, but the file size is the same for the same track.
![image](https://user-images.githubusercontent.com/1609513/218600978-b0b946ce-38a1-4e96-9f0e-ea59a7a62a06.png)

**Comment by junh1024** [comment]
> *Full object support (?): I'm told that TrueHD stores the clustered objects separately (so lossy clustering->lossless (no JOC) ->lossless encode), unlike EAC3 where it's combined into 5.1/7.1 channels using JOC (so lossy clustering->lossy JOC->lossy encode), but mediainfo and ffmpeg all give me 7.1 channels for TrueHD, so I'm not sure what's happening. Maybe TrueHD is already fully supported by FFmpeg (and therefore most of the players).

For both THDA & DDPA, the objects are effectively joined due to the Legacy compatibility layer. For THDA,  the object reconstruction process is lossless because the codec itself is lossless.

**Comment by sclsj** [comment]
> You have to demux from tidal in order to play in Dolby Reference Player.
> 
> > > > I can't find a player for AC4 or TrueHD*. I'm still in the process of requesting a trial to the Dolby Reference Player but I haven't heard back for a while (2+ weeks) so it's probably ignored.
> > > > Or will this project be for eac3 only?
> > > > *Full object support (?): I'm told that TrueHD stores the clustered objects separately (so lossy clustering->lossless (no JOC) ->lossless encode), unlike EAC3 where it's combined into 5.1/7.1 channels using JOC (so lossy clustering->lossy JOC->lossy encode), but mediainfo and ffmpeg all give me 7.1 channels for TrueHD, so I'm not sure what's happening. Maybe TrueHD is already fully supported by FFmpeg (and therefore most of the players).
> > > 
> > > 
> > > Dolby Reference Player only supports AC4-IMS playback to virtualized headphones and virtualized stereo speakers as experienced on Android devices supporting Atmos. AC4-IMS is used by Tidal and according to Dolby Apple Music for Android users.
> > 
> > 
> > Tidal on iOS seems to not be using system Dolby Atmos decoder because the control center shows "stereo". How can I tell if it's actually playing the Dolby Atmos version, and if it is, if it's streaming eac3 or ac4?
> 
> Apple devices don't support ac4-ims, only dd+joc so Tidal streams dd+joc to apple and ac4-ims to Android. Dolby say Apple Music stream ac4-ims to Android, but I received dd+joc when I looked into it. Looks like Apple Music encodes the joc differently, but the file size is the same for the same track. Top is Tidal, bottom is Apple Music. ![image](https://user-images.githubusercontent.com/1609513/218600978-b0b946ce-38a1-4e96-9f0e-ea59a7a62a06.png)

How did you capture the dd+joc from Apple Music? I have long wanted to compare the stereo and spatial versions of some tracks but they are only on Apple Music and not tidal.

"Tidal streams dd+joc to apple" - does Tidal uses its own decoder from Dolby or the system decoder on iOS? I believe it's the first case but I can't be sure.

**Comment by ValZapod** [comment]
>believe it's the first case but I can't tell.

Just as on Android it uses OS private blob system decoder, of course.

**Comment by hockinsk** [comment]
Unfortunately I haven't been able to repeat it, but essentially I installed Apple Music on Android, downloaded within the app the Fleetwood Mac track and then copied it across USB to my PC. Essentially, the Apple Music e-ac3 and Tidal e-ac3 are identical however. iirc I was messing around with Apple Music saving to SD Card or Internal memory and just got lucky it left a file there that I downloaded from Apple Music.
